src/query/pipeline/transforms/src/processors/transforms/transform_k_way_merge_sort.rs

```diff
@@ -0,0 +1,817 @@
+        let input = &self.inputs[i];
+        let buffer = &mut self.buffer[i];
+        let info = &mut self.info[i];
+
+        const TASK_POS: usize = 2;
+        const TASK_ROWS_POS: usize = 1;
+
+        if let Some(info) = info {
+            if info.done() {
+                return if info.task == self.cur_task {
+                    Ok(PullEvent::Data)
+                } else {
+                    Ok(PullEvent::Pending)
+                };
+            }
+        }
+
+        if input.has_data() {
+            let mut block = input.pull_data().unwrap()?;
+            input.set_need_data();
+
+            let task = match info {
+                None => {
+                    let task = get_u32(&block, TASK_POS);
+                    let task_rows = get_u32(&block, TASK_ROWS_POS);
+                    debug_assert!(task_rows > 0);
+                    debug_assert!(
+                        task >= self.cur_task,
+                        "disorder task. cur_task: {} task: {task} task_rows: {task_rows}",
+                        self.cur_task
+                    );
+                    let remain = task_rows as usize - block.num_rows();
+                    let _ = info.insert(Info {
+                        task,
+                        total: task_rows as usize,
+                        remain,
+                    });
+                    block.pop_columns(2);
+                    buffer.push_back(block);
+                    task
+                }
+                Some(info) => {
+                    debug_assert_eq!(info.task, get_u32(&block, TASK_POS));
+                    debug_assert_eq!(info.total, get_u32(&block, TASK_ROWS_POS) as usize);
+
+                    let rows = block.num_rows();
+                    if rows <= info.remain {
+                        block.pop_columns(2);
+                        buffer.push_back(block);
+                        info.remain -= rows;
+                        info.task
+                    } else {
+                        unreachable!()
+                    }
+                }
+            };
+            return if task == self.cur_task {
+                Ok(PullEvent::Data)
+            } else {
+                Ok(PullEvent::Pending)
+            };
+        }
+
+        if !buffer.is_empty() {
+            return if info.unwrap().task == self.cur_task {
+                Ok(PullEvent::Data)
+            } else {
+                Ok(PullEvent::Pending)
+            };
+        }
+
+        if input.is_finished() {
+            return Ok(PullEvent::Finished);
+        }
+
+        input.set_need_data();
+        Ok(PullEvent::Pending)
+    }
+
+    fn push(&mut self) -> bool {
+        let cur_input = self.cur_input.unwrap();
+        let buffer = &mut self.buffer[cur_input];
+        let info = &mut self.info[cur_input];
+        debug_assert_eq!(info.unwrap().task, self.cur_task);
+        if let Some(block) = buffer.pop_front() {
```

Comment:
block may be very small, we need to concat blocks in buffer with MAX_BLOCK_SIZE

Reply:
> block may be very small, we need to concat blocks in buffer with MAX_BLOCK_SIZE

depend on https://github.com/datafuselabs/databend/pull/16401